### PR TITLE
FOGL-8461 Add pipeline splitting

### PR DESCRIPTION
**Pull request by MarkRiddoch**

### CODE REVIEW

File: C/common/filter_pipeline.cpp [hunk 1]
```diff
@@ -248,6 +175,43 @@ bool FilterPipeline::loadFilters(const string& categoryName)
...
+			pipeline.emplace_back(element);
+		}
+		else if (itr->IsObject())
+		{
+			// An object, probably the write destination
+			Logger::getLogger()->warn("This version of Fledge does not support pipelines with different destinations. The destination will be ignored and the data written to the default storage service.");
+		}
+		else
+		{
+			Logger::getLogger()->error("Unexpected object in  pipeline definition, ignoring");
```
Comment (RVerhoeff) [inline C/common/filter_pipeline.cpp:210]: Extra blank between "in" and "pipeline."

File: C/common/filter_pipeline.cpp [hunk 2]
```diff
@@ -410,27 +321,53 @@ void FilterPipeline::cleanupFilters(const string& categoryName)
...
 	auto it = m_filterCategories.find(category);
 	if (it != m_filterCategories.end())
 	{
 		it->second->reconfigure(newConfig);
 	}
 }
 
+/**
+ * Called when we pass the data into the pipeline. Set the
+ * numebr of active branches to 1
```
Comment (RVerhoeff) [inline C/common/filter_pipeline.cpp:333]: Typo: numebr -> number

File: C/common/pipeline_branch.cpp [hunk 3]
```diff
@@ -0,0 +1,268 @@
...
+ * If the pipeline is not already shutdown then shut it down
+ * Delete the thread if it exists.
+ */
+PipelineBranch::~PipelineBranch()
+{
+	if (!m_shutdownCalled)
+	{
+		m_shutdownCalled = true;
+		m_cv.notify_all();
+		m_thread->join();
```
Comment (RVerhoeff) [inline C/common/pipeline_branch.cpp:39]: Is it necessary to call _m_thread->joinable()_ before doing this?

File: C/common/pipeline_branch.cpp [hunk 4]
```diff
@@ -0,0 +1,268 @@
...
+	for (auto it = m_branch.begin(); it != m_branch.end(); ++it)
+	{
+		delete *it;
+	}
+}
+
+/**
+ * Setup the configuration for a branch in a pipeline
+ *
+ * @param       mgtClient       The managament client
```
Comment (RVerhoeff) [inline C/common/pipeline_branch.cpp:62]: Typo: managament -> management

File: C/common/pipeline_branch.cpp [hunk 5]
```diff
@@ -0,0 +1,268 @@
...
+void PipelineBranch::shutdown(ServiceHandler *serviceHandler, ConfigHandler *configHandler)
+{
+	// Shutdown the handler thread
+	m_shutdownCalled = true;
+	m_cv.notify_all();
+	m_thread->join();
+	delete m_thread;
+	m_thread = NULL;
+
+	// Shutdown the fitler elements on the branch
```
Comment (RVerhoeff) [inline C/common/pipeline_branch.cpp:209]: Typo: fitler -> filter

File: C/common/pipeline_branch.cpp [hunk 6]
```diff
@@ -0,0 +1,268 @@
...
+	while (!m_queue.empty())
+	{
+		ReadingSet *readings = m_queue.front();
+		m_queue.pop();
+		delete readings;
+	}
+}
+
+/**
+ * Return if the brach is ready to be executed
```
Comment (RVerhoeff) [inline C/common/pipeline_branch.cpp:225]: Typo: brach -> branch

File: C/common/pipeline_filter.cpp [hunk 7]
```diff
@@ -0,0 +1,215 @@
...
+		// Failure
+		return NULL;
+	}
+	Logger::getLogger()->info("Loading filter plugin '%s'.", filterName.c_str());
+
+	PluginManager *manager = PluginManager::getInstance();
+	PLUGIN_HANDLE handle;
+	if ((handle = manager->loadPlugin(filterName, PLUGIN_TYPE_FILTER)) != NULL)
+	{
+		// Suceess
```
Comment (RVerhoeff) [inline C/common/pipeline_filter.cpp:113]: Typo: Suceess -> Success

File: C/services/south/ingest.cpp [hunk 8]
```diff
@@ -744,12 +744,13 @@ void Ingest::processQueue()
 									  "filter pipeline is ready");
 						std::this_thread::sleep_for(std::chrono::milliseconds(150));
 					}
-
 					ReadingSet *readingSet = new ReadingSet(m_data);
 					m_data->clear();
+					m_filterPipeline->execute();	// Set the pipelien executing
```
Comment (RVerhoeff) [inline C/services/south/ingest.cpp:749]: Typo: pipelien -> pipeline

File: C/services/south/ingest.cpp [hunk 9]
```diff
@@ -1076,9 +1080,14 @@ void Ingest::useFilteredData(OUTPUT_HANDLE *outHandle,
 	    Logger::getLogger()->info("%s:%d: Input readingSet modified by filter: ingest->m_data=%p, readingSet->getAllReadingsPtr()=%p",
```
Comment (RVerhoeff) [inline C/services/south/ingest.cpp:1080]: Should this be a _debug_ message instead of an _info_?

File: Makefile [hunk 10]
```diff
@@ -25,7 +25,7 @@ ifneq ("$(PLATFORM_RH)","")
 else
 	PIP_INSTALL_REQUIREMENTS := python3 -m pip install -Ir
 	PYTHON_BUILD_PACKAGE = python3 setup.py build -b ../$(PYTHON_BUILD_DIR)
-	CMAKE := cmake
+	CMAKE := cmake -DCMAKE_BUILD_TYPE=Debug
```
Comment (RVerhoeff) [inline Makefile:28]: Did you intend to build for Debug in all cases?

**Review by RVerhoeff — CHANGES_REQUESTED** [review]
Mostly typos plus a few other items

**Review by pintomax — APPROVED** [review]
LGTM,

just minor fixes to implement as already reported

File: C/common/pipeline_branch.cpp (hunk 3 above)
Comment (MarkRiddoch) [reply C/common/pipeline_branch.cpp:39]: It should not be, but it is a good precaution to take

File: C/services/south/ingest.cpp (hunk 9 above)
Comment (MarkRiddoch) [reply C/services/south/ingest.cpp:1080]: The code should not be three at all as if was #if 0

**Review by RVerhoeff — APPROVED** [review]

**Pull request merged**